### DISCUSSION TITLE
Handling Mixed Second-Order Partial Derivatives in MOOSE

### DISCUSSION THREAD

**Discussion by hityyds** [idaholab/moose#28246] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hi everyone,

I need to solve an equation that includes a mixed second-order partial derivative with respect to x and y, but I'm unsure how to handle the mixed partial derivative in MOOSE. In 2021, a similar question was asked (#https://github.com/idaholab/moose/discussions/19475), and the recommendation was to use integral transforms for differential reduction. I've tried this method, but MOOSE fails to converge during the computation.

I'm curious if there have been any updates or advancements in how MOOSE handles second-order mixed partial derivatives. I noticed that the code related to the incompressible Navier-Stokes (INS) equations in MOOSE seems to include variables for second-order derivatives (e.g., `_second_u_vel(coupledSecond("u"))`), although I recall that the INS module does not include mixed second-order partial derivatives.

The specific mixed second-order partial derivative I need to handle is:
![20240728224503](https://github.com/user-attachments/assets/59db9f95-624b-4108-af28-4f74efefbf8e)

Any guidance or insights on how to properly implement this in MOOSE would be greatly appreciated. Thank you!

**Comment by GiudGiud** [comment]
Hello

I think coupledSecond returns a tensor and you can simply access the off-diagonal component to get the d2f/dxdy 
I have never tried but I would expect that to be the solution.
You should make a simple auxkernel with coupledSecond and use it on predefined second order variables, for which you know the correct 2nd order derivatives, to check that it works


Guillaume

**Reply by hityyds** [reply]
Thank you very much for your response. I also believe this approach should work, but I am unsure how to access the off-diagonal elements of the tensor. I found the definition of `VariableSecond` in the code, but I am not familiar with the corresponding class. Do you know anything about this? Here is some code I found regarding the definition of `VariableSecond`:
```
// Seconds, this is defined in INSBase and _second_u_vel is a data member.
  const VariableSecond & _second_u_vel;

// This is declared in framework/include/MooseTypes.h
typedef typename OutputTools<Real>::VariableSecond VariableSecond;

// This is also declared in framework/include/MooseTypes.h
template <typename OutputType>
struct OutputTools
{
  typedef typename TensorTools::IncrementRank<OutputType>::type OutputGradient;
  typedef typename TensorTools::IncrementRank<OutputGradient>::type OutputSecond;
  ......
  typedef MooseArray<OutputSecond> VariableSecond;
  ......
}
```
In INSBase, I found the following operations:
```
_second_u_vel[_qp].tr();
_second_u_vel[_qp].row(0);
```

**Reply by GiudGiud** [reply, marked as answer]
I believe you can output the tensor with 

std::cout << Moose::stringify(..

If you do _tensor.row(0)(1) you should get the xy mixed derivative

Note that you also need to index by quadrature point with [_qp]

**Reply by hityyds** [reply]
Thank you, it works.
